examples/window-app/nrfconnect/main/WindowCovering.cpp

```diff
@@ -0,0 +1,297 @@
+/*
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "WindowCovering.h"
+#include "AppConfig.h"
+#include "PWMDevice.h"
+
+#include <dk_buttons_and_leds.h>
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app/util/af.h>
+#include <logging/log.h>
+#include <platform/CHIPDeviceLayer.h>
+#include <zephyr.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_MATTER_LOG_LEVEL);
+
+using namespace ::chip::Credentials;
+using namespace ::chip::DeviceLayer;
+using namespace chip::app::Clusters::WindowCovering;
+
+static k_timer sMoveTimer;
+static constexpr uint32_t sMoveTimeoutMs{ 200 };
+
+static constexpr float FromOneRangeToAnother(uint32_t aInMin, uint32_t aInMax, uint32_t aOutMin, uint32_t aOutMax, uint32_t aInput)
+{
+    const auto diffInput  = aInMax - aInMin;
+    const auto diffOutput = aOutMax - aOutMin;
+    if (diffInput > 0)
+    {
+        return aOutMin + ((float) diffOutput / (float) diffInput) * (aInput - aInMin);
```

Comment:
```suggestion
        return static_cast<uint32_t>(aOutMin + static_cast<uint64_t>(aInput - aInMin) * diffOutput / diffInput);
```
By swapping some components you can avoid using floats (and be able to compile the code without FPU).

Reply:
Are you sure that the `diffOutput / diffInput` will not yield the integer type with value of 0 every time diffInput > diffOutput, and the whole expression will be equal to 0?

Reply:
If only `static_cast<uint64_t>(aInput - aInMin) * diffOutput > diffInput`, no :)